src/finn/transformation/merge_onnx_models.py

```diff
@@ -0,0 +1,171 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import copy
+
+from onnx import helper
+
+from finn.transformation import Transformation
+from finn.core.modelwrapper import ModelWrapper
+import finn.util.basic as util
+from finn.transformation.infer_shapes import InferShapes
+from finn.transformation.infer_datatypes import InferDataTypes
+from finn.transformation.general import (
+    GiveReadableTensorNames,
+    GiveUniqueNodeNames,
+    GiveUniqueParameterTensors,
+)
+
+
+def _make_model_values_unique(model1, model2):
+    # ensure that tensor and node names are different in each model
+    # tensors
+    names1 = model1.get_all_tensor_names()
+    names2 = model2.get_all_tensor_names()
+    duplicates = list(set(names1).intersection(names2))
+    # if there are duplicates in the tensor names rename these tensors
+    if duplicates:
+        used_names = names1 + names2
+        for name in duplicates:
+            # model1
+            new_name = util.random_string()
+            while new_name in used_names:
+                new_name = util.random_string()
+            model1.rename_tensor(name, new_name)
+            used_names.append(new_name)
+
+            # model2
+            new_name = util.random_string()
+            while new_name in used_names:
+                new_name = util.random_string()
+            model2.rename_tensor(name, new_name)
+            used_names.append(new_name)
+
+    # nodes
+    names1 = [x.name for x in model1.graph.node]
+    names1 = list(filter(None, names1))  # filter out empty node names
+    names2 = [x.name for x in model2.graph.node]
+    names2 = list(filter(None, names2))
+    duplicates = list(set(names1).intersection(names2))
+    # if there are duplicates erase all node names
+    if duplicates:
+        for n in model1.graph.node:
+            n.name = ""
+        for n in model2.graph.node:
+            n.name = ""
+
+    return (model1, model2)
+
+
+class MergeONNXModels(Transformation):
+    """Merges two models. The model passed in the transformation will be inserted before
+    the model the transformation is applied on, the resulting model is returned."""
+
+    def __init__(self, pre_model):
+        super().__init__()
+        # use deep copy of model that should be inserted in the beginning of
+        # the other model to ensure that it stays unchanged
+        self.pre_model = copy.deepcopy(pre_model)
+
+    def apply(self, model):
+        graph_modified = False
+        pre_model = self.pre_model
+
+        # make model values unique to avoid any conflict
+        (pre_model, model) = _make_model_values_unique(pre_model, model)
+
+        # check if models can be merged
+        output_model_a = pre_model.graph.output[0].name
+        input_model_b = model.graph.input[0].name
```

Comment:
this will only connect 1 output of the pre graph to 1 input of the post graph. what if they have more outputs/inputs? this limitation should be specified in the transformation description.

you may also want to check num dynamic outputs in pre == num dynamic inputs in post. dynamic means that tensor has no initializer associated with it.